### DISCUSSION TITLE
Cahn-Hilliard not being coservative

### DISCUSSION THREAD

**Discussion by alimostafavi24** [idaholab/moose#27346] [Q&A Modules: Phase field]
Dear Moose Community,

It is well known that Cahn-Hilliard equation is in conservative form, and it can be shown mathematically that with Neumann BC for the chemical potential, the integral of order parameter over the domain must be constant during time.

I noticed that within the Moose framework, this is not true on adaptive grids. What will be the best explanation for this behavior? Is there any good reference for this issue on adaptive grids?

Thanks,
Ali

**Comment by GiudGiud** [comment]
Hello

Is the order parameter one of the variables of the solve?

When we perform steps of adaptivity, whether refinement or coarsening, we project the variables onto the new grid. Nonlinear quantities related to these variables would not be conserved by this process. 

You could create a user object that enforces this conservation. 

Guillaume

**Reply by alimostafavi24** [reply]
Thank you for your reply.

Yes, I perform adaptivity on the order parameter c and the conservation is lost. 

I solve the issue of losing the convergence in another way. Right now I am mainly looking for a way to describe why exactly conservation is lost on adaptive grids (is it related to the interpolation of the variables to the new grid?) and whether this is a common issue in finite element method with h-refinement.

Also, does this happen to all of the non-linear variables in the system? (5 other variables apart from order parameter including velocity components, pressure, chemical potential)

And if there is any reference that I can cite and refer to about this issue it is highly appreciated.

Thanks.

**Reply by GiudGiud** [reply]
Is 'c' one of the nonlinear variables?
What finite element family and order is it?

**Reply by alimostafavi24** [reply]
Yes, it is.

```
Nonlinear System:
  Num DOFs:                115298
  Num Local DOFs:          9940
  Variables:               { "vel_x" "vel_y" "p" "c" "w" "rho_e" "phi" }
  Finite Element Types:    "LAGRANGE"  
  Approximation Orders:    "FIRST"
```

**Reply by GiudGiud** [reply]
So when refining, we should easily conserve the integral of c.

When coarsening, I would expect us to minimize the L2 difference but we ll have to check libmesh to be sure

**Reply by alimostafavi24** [reply]
I think you are right. I am adding this MOOSE input file for test purposes.

When there is no mesh coarsening and just h-refinement, it seems that solution is conserved. But when coarsening is enabled (set it more than 0.0 in the adaptivity part of the input file), then the solution is not conservative.


```
[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 50
  ny = 50
  nz = 0
  xmin = 0
  xmax = 250
  ymin = 0
  ymax = 250
  zmin = 0
  zmax = 0
  elem_type = QUAD4
[]

[Variables]
  [./c]
  [../]
  [./w]
  [../]
[]

[AuxVariables]
  [./local_free_energy]
    order = CONSTANT
    family = MONOMIAL
  [../]
[]

[ICs]
  [./cIC]
    type = SmoothCircleIC
    variable = c
    x1 = 125.0
    y1 = 125.0
    radius = 60.0
    invalue = 1.0
    outvalue = 0.1
    int_width = 30.0
  [../]
[]

[Kernels]
  [./c_res]
    type = SplitCHParsed
    variable = c
    f_name = F
    kappa_name = kappa_c
    w = w
  [../]
  [./w_res]
    type = SplitCHWRes
    variable = w
    mob_name = M
  [../]
  [./time]
    type = CoupledTimeDerivative
    variable = w
    v = c
  [../]
[]

[AuxKernels]
  [./local_free_energy]
    type = TotalFreeEnergy
    variable = local_free_energy
    kappa_names = kappa_c
    interfacial_vars = c
  [../]
[]

[Materials]
  [./pfmobility]
    type = GenericConstantMaterial
    prop_names  = 'M kappa_c'
    prop_values = '1e-3 0.1'
  [../]
  [./free_energy]
    type = DerivativeParsedMaterial
    coupled_variables = c
    constant_names = 'barr_height  cv_eq'
    constant_expressions = '0.1          1.0e-2'
    expression = 16*barr_height*(c-cv_eq)^2*(1-cv_eq-c)^2
    derivative_order = 2
  [../]
[]

[Postprocessors]

  [./total_free_energy]
    type = ElementIntegralVariablePostprocessor
    variable = local_free_energy
  [../]

    [./total_solutec]
    type = ElementIntegralVariablePostprocessor
    variable = c
    execute_on = linear

  [../]


[]

[Preconditioning]
  [./SMP]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient
  scheme = bdf2
  solve_type = NEWTON
  petsc_options_iname = -pc_type
  petsc_options_value = lu
  l_max_its = 30
  l_tol = 1.0e-4
  nl_rel_tol = 1.0e-10
  start_time = 0.0
  num_steps = 60
  dt = 200
[]

[Outputs]
  execute_on = 'timestep_end'
  exodus = true
[]

[Adaptivity]
  initial_steps = 2
  steps = 1
  marker = error
  [Indicators]
    [jump]
      type = GradientJumpIndicator
      variable = c
     outputs = none

    []
  []
  [Markers]
    [error]
      type = ErrorFractionMarker
     coarsen = 0.50
      refine = 0.90
      indicator = jump
      outputs = none
    []
  []
 max_h_level = 2
[]
```

**Reply by GiudGiud** [reply]
@roystgnr could you please point us to the docs for h-coarsening of variables in libmesh?
I see the code for coarsening elements in MeshRefinement. Is the code for the projection during coarsening in the GenericProjector?

**Reply by alimostafavi24** [reply]
This input file from MOOSE tests shows it better.

```
[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 100
  ny = 100
  xmax = 30.0
  ymax = 30.0
  elem_type = QUAD4
[]

[Variables]
  [./c]
  [../]
  [./w]
  [../]
[]

[ICs]
  [./c_IC]
    type = CrossIC
    x1 = 0.0
    x2 = 30.0
    y1 = 0.0
    y2 = 30.0
    variable = c
  [../]
[]

[Kernels]
  [./cres]
    type = SplitCHParsed
    variable = c
    kappa_name = kappa_c
    w = w
    f_name = F
  [../]
  [./wres]
    type = SplitCHWRes
    variable = w
    mob_name = M
    coupled_variables = c
  [../]
  [./time]
    type = CoupledTimeDerivative
    variable = w
    v = c
  [../]
[]

[BCs]
  [./Periodic]
    [./all]
      auto_direction = 'x y'
    [../]
  [../]
[]

[Materials]
  [./kappa]
    type = GenericConstantMaterial
    prop_names = 'kappa_c'
    prop_values = '2e-2'
  [../]
  [./mob]
    type = DerivativeParsedMaterial
    property_name = M
    coupled_variables = c
    expression = '1-0.9*c^2'
    outputs = exodus
    derivative_order = 1
  [../]
  [./free_energy]
    type = MathEBFreeEnergy
    property_name = F
    c = c
  [../]
[]

[Preconditioning]
  [./SMP]
   type = SMP
   off_diag_row = 'w c'
   off_diag_column = 'c w'
  [../]
[]

[Executioner]
  type = Transient
  scheme = 'BDF2'

  solve_type = 'NEWTON'
  petsc_options_iname = '-pc_type -ksp_gmres_restart -sub_pc_type -pc_asm_overlap'
  petsc_options_value = 'asm         31      lu      1'

  l_max_its = 30
  l_tol = 1.0e-4
  nl_max_its = 50
  nl_rel_tol = 1.0e-10

  dt = 1.0
  num_steps = 20
[]

[Outputs]
  exodus = true
[]

[Postprocessors]


  # for testing we also monitor the total solute amount, which should be conserved.


    [./total_solutec]
    type = ElementIntegralVariablePostprocessor
    variable = c

  [../]
  []

  [Adaptivity]
  initial_steps = 1
  steps = 1
  marker = error
  [Indicators]
    [jump]
      type = GradientJumpIndicator
      variable = c
     outputs = none

    []
  []
  [Markers]
    [error]
      type = ErrorFractionMarker
     coarsen = 0.90
      refine = 0.90
      indicator = jump
      outputs = none
    []
  []
 max_h_level = 1
[]
```

**Reply by roystgnr** [reply]
It's GenericProjector that does the refinement and coarsening.  In either case the algorithm is the same: interpolate values (and for C1 elements gradients, and for HERMITE elements gradient cross terms) at nodes, then project (L2 for C0 elements, H1 for C1) along edges holding nodes fixed, then project along faces holding edges and nodes fixed, then project on interiors holding boundaries fixed.

For refinement, this gives you the same value (up to floating point error) for p-refinement and for h-refinement in any FEFamily that has a nested space (everything except CLOUGH, in libMesh).

For coarsening, this totally breaks conservation.  Picture the simplest case, 1D p=1.  You have a coarse element that's linear between its two end points, you set its endpoints equal to what the two fine elements had ... and now you're done; there's no other DoFs to set.  If the coarsened-away node's value was higher than the average of its neighbors then you've just subtracted mass to nowhere, if lower you've just added mass from nowhere.

The only alternatives are non-local (as is fairly easy to see in this case - you can't tweak those end node values without worrying about how it affects conservation on their neighbors, and how to compensate for that without breaking *their* neighbors...). E.g. doing a global L2 projection rather than a bunch of local ones would be conservative ... but the problem in *that* case is that you can more easily get something like Gibbs' ringing effects; there's no free lunch once you start shrinking your approximation space.

**Reply by alimostafavi24** [reply]
Thanks. It was explained very well :)

**Reply by alimostafavi24** [reply]
Is there any reference that I can cite or refer to for this numerical problem? @roystgnr 

Thanks,

**Reply by roystgnr** [reply]
If there's no such reference then I'm going to start kicking myself; I'd be a jackass to notice something like that, publish papers on related topics, and yet never mention the conservation issues.

... and, looking through them, it looks like I'm a jackass.  My dissertation comes the closest, when I mention that, even when using a formulation that guarantees a monotonic-decreasing global-free-energy-functional value, adaptive coarsening can lead to small local increases ... but I never mention the effect on conservation (which is simpler to understand and just as important), nor the effect on quadrature-based integration for chemical-free-energy functionals with bounded domains (which is a little harder to understand but is even more important).  Gah.  Up until this day I was most embarrassed about misusing the word "utilize" in my dissertation; this really blows that out of the water.

In my defense, I'm looking at adaptivity textbooks right now and also not finding any mention of the problem there, and a (too-brief) literature search isn't helping either.  Surely people didn't just think this was too obvious to mention?  Or maybe most adaptivity people are either already using non-conservative formulations (this seems to be common in the papers I'm seeing...) and don't care about the problem or are using DG (many versions of which also give you conservative local coarsening operators) and don't have the problem?

**Reply by alimostafavi24** [reply]
Thanks for the clarification.

I'm using coupled Cahn-Hilliard-Navier-Stokes equations to simulate two-phase flows. I need to use very small interface thickness, so I definitely need mesh refinement and coarsening at the same time.

I can't really have global conservation issue. For example, during droplet's motion, it's area/volume constantly decreases and finally vanishes. I could get results that look good initially, but exacerbated during the simulation. I mean the whole results were questionable when mass loss was present. But I finally resolved it with other techniques.

It is quite suprising that the issue of the solution of the Cahn-Hilliard equation not being conservative while coarsening the grid has not been addressed anywhere to the best of my knowledge. Many people use it and they might utilize it on adaptive grids too, but never pointed out to this fact. I don't know yet to do it or not if I am 100% correct :)

**Reply by roystgnr** [reply]
>For example, during droplet's motion, it's area/volume constantly decreases and finally vanishes.

This is actually not a good example!  If you have multiple droplets in a Cahn-Hilliard simulation, the very-long-term behavior can actually be for a smaller droplet to "evaporate", even when there is global mass conservation, because the mass ends up traveling through the not-quite-uniform c≈0 part of the domain to a larger droplet.  This is a very slow process, I'd guess the time it takes is exponential in the ratio of distance-between-droplets/equilibrium-layer-thickness, but it happens.  Definitely check the global mass integral to see how far off from conservation you are; don't assume weird local behavior must indicate something's wrong.

I'd actually even check the global mass integral in a uniform simulation, just to make sure everything's solid with the formulation.  I vaguely remember looking at CHNS once, and combining those properly even before you get to the discretization level was surprisingly more complicated than I'd expected.

If everything is solid, you might look at DG formulations.  I personally liked using C1 elements rather than pairs of C0 elements for CH, and going past C0 down to DG feels even crazier, but that would at least make it easy to eliminate mass conservation issues when coarsening.  The only other good option I can see would be using elements with "bubble functions" (e.g. p=2 if you're using C0 elements) so you can tweak the bubble function coefficient for mass correction after a coarsening step.  That would require new C++ code, though; no getting it out of a simple MOOSE input file.
